README.rst
Flake8 Coding plugin
=====================

.. image:: https://travis-ci.org/tk0miya/flake8-coding.svg?branch=master
   :target: https://travis-ci.org/tk0miya/flake8-coding

.. image:: https://coveralls.io/repos/tk0miya/flake8-coding/badge.svg?branch=master
   :target: https://coveralls.io/r/tk0miya/flake8-coding?branch=master

.. image:: https://codeclimate.com/github/tk0miya/flake8-coding/badges/gpa.svg
   :target: https://codeclimate.com/github/tk0miya/flake8-coding
   :alt: Code Climate


Adds coding magic comment checks (``coding:``) to ``flake8``.

Install
--------

Install with pip::

    $ pip install flake8-coding

You can check that ``flake8`` has picked it up by looking for ``flake8_coding``
in the output of ``--version``:

.. code-block:: sh

    $ flake8 --version
    2.5.4 (pep8: 1.7.0, pyflakes: 1.0.0, flake8_coding: 1.1.1, mccabe: 0.4.0) CPython 2.7.11 on Darwin

Options
-------

``accept-encodings``
~~~~~~~~~~~~~~~~~~~~

A comma-separated list of acceptable source code encodings for the ``coding:``
magic comments in files. Default is ``latin-1, utf-8``.

You can pass this as a command-line argument to ``flake8``, e.g.
``--accept-encodings=utf-8,utf-16``, or put it in your config file, e.g.:

.. code-block:: ini

    [flake8]
    accept-encodings = utf-8,utf-16

``no-accept-encodings``
~~~~~~~~~~~~~~~~~~~~~~~

If activated, this disallows all ``coding:`` magic comments, no matter their
encoding. This might be useful for Python 3 projects where UTF-8 is the default
and you don't want other encodings used in your project.

You can pass this as a command-line argument to ``flake8``, e.g.
``--no-accept-encodings``, or put it in your config file, e.g.:

.. code-block:: ini

    [flake8]
    no-accept-encodings = True

Rules
-----

C101 Coding magic comment not found
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

No magic encoding comment was found in the file. As per
`PEP-263 <https://www.python.org/dev/peps/pep-0263/>`_, this must be in the
first two lines of the file.

C102 Unknown encoding found in coding magic comment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The encoding found in the magic encoding comment did not match the
``accept-encodings`` option.

C103 Coding magic comment present
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``no-accept-encodings`` is set, and a magic encoding comment was found in the
file.

Requirements
-------------

* Python 2.7, 3.4
* flake8

License
--------

Apache License 2.0


CHANGES.rst
Changelog
=========

1.3.2 (2019-06-16)
------------------
* Fix a bug

  - #18 Use first line number for C101 message

1.3.1 (2016-09-09)
------------------
* Fix bugs

  - #14 Can't read from stdin using flake8-coding and flake8-print at the same time
  - #16 flake8-coding fails on python 3.7 with "RuntimeError: generator raised StopIteration"

1.3.0 (2016-08-20)
------------------
* Drop Python 2.6 support
* Fix a bug

  - Fix no-accept-encodings to work in config files

1.2.2 (2016-06-28)
------------------
* Fix a bug

  - Could not work under flake8 >= 3.0

1.2.1 (2016-05-20)
------------------
* Add option `no-accept-encodings`. If set, will warn for files containing a `coding:` magic comment.
* Fix a bug

  - #4 stdin not supported

1.2.0 (not released)
--------------------

1.1.1 (2015-10-25)
------------------
* Fix a bug

  - Fix #2 ignore errors if target file is not found

1.1.0 (2015-06-27)
------------------
* Do not warn for empty file (Thanks to gforcada)

1.0.1 (2015-03-25)
------------------
* Fix a bug

  - Fix typo in option name: accept-encodings

1.0.0 (2015-03-21)
------------------
* Initial release


run_tests.py
# -*- coding: utf-8 -*-

import unittest
from collections import namedtuple

from flake8_coding import CodingChecker

from mock import patch


Options = namedtuple('Options', 'no_accept_encodings, accept_encodings')


class TestFlake8Coding(unittest.TestCase):
    def test_has_utf8_coding_header(self):
        checker = CodingChecker(None, 'testsuite/utf8.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_file_not_found(self):
        checker = CodingChecker(None, 'file_not_found')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_empty_file(self):
        checker = CodingChecker(None, 'testsuite/empty.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_has_latin1_coding_header(self):
        checker = CodingChecker(None, 'testsuite/latin1.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_has_coding_header_at_2nd_line(self):
        checker = CodingChecker(None, 'testsuite/2nd-line.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_has_coding_header_at_3rd_line(self):
        checker = CodingChecker(None, 'testsuite/3rd-line.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(len(ret), 1)
        self.assertEqual(ret[0][0], 1)
        self.assertEqual(ret[0][1], 0)
        self.assertTrue(ret[0][2].startswith('C101 '))

    def test_has_vim_styled_coding_header(self):
        checker = CodingChecker(None, 'testsuite/vim-style.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_has_coding_header_with_invalid_encoding_name(self):
        checker = CodingChecker(None, 'testsuite/invalid.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(len(ret), 1)
        self.assertEqual(ret[0][0], 2)
        self.assertEqual(ret[0][1], 0)
        self.assertTrue(ret[0][2].startswith('C102 '))

    def test_has_no_coding_headers(self):
        checker = CodingChecker(None, 'testsuite/nocodings.py')
        checker.encodings = ['latin-1', 'utf-8']
        ret = list(checker.run())
        self.assertEqual(len(ret), 1)
        self.assertEqual(ret[0][0], 1)
        self.assertEqual(ret[0][1], 0)
        self.assertTrue(ret[0][2].startswith('C101 '))

    def test_default_encoding(self):
        try:
            options = Options(False, 'latin-1, utf-8')
            CodingChecker.parse_options(options)
            self.assertEqual(CodingChecker.encodings, ['latin-1', 'utf-8'])
        finally:
            if hasattr(CodingChecker, 'encodings'):
                del CodingChecker.encodings

    def test_change_encoding(self):
        try:
            options = Options(False, 'utf-8,utf-16')
            CodingChecker.parse_options(options)
            self.assertEqual(CodingChecker.encodings, ['utf-8', 'utf-16'])
        finally:
            if hasattr(CodingChecker, 'encodings'):
                del CodingChecker.encodings

    def test_stdin(self):
        try:
            from flake8.engine import pep8 as stdin_utils  # noqa
            target = 'flake8.engine.pep8.stdin_get_value'
        except ImportError:
            from flake8 import utils as stdin_utils # noqa
            target = 'flake8.utils.stdin_get_value'

        with patch(target) as stdin_get_value:
            with open('testsuite/nocodings.py') as fp:
                stdin_get_value.return_value = fp.read()

            for input in ['stdin', '-', None]:
                checker = CodingChecker(None, input)
                checker.encodings = ['latin-1', 'utf-8']
                ret = list(checker.run())
                self.assertEqual(len(ret), 1)
                self.assertEqual(ret[0][0], 1)
                self.assertEqual(ret[0][1], 0)
                self.assertTrue(ret[0][2].startswith('C101 '))

    def test_no_accept_encodings_sets_encodings_none(self):
        try:
            options = Options(True, 'latin-1,utf-8')
            CodingChecker.parse_options(options)
            self.assertTrue(CodingChecker.encodings is None)
        finally:
            if hasattr(CodingChecker, 'encodings'):
                del CodingChecker.encodings

    def test_encoding_none_with_no_coding_comment(self):
        checker = CodingChecker(None, 'testsuite/nocoding.py')
        checker.encodings = None
        ret = list(checker.run())
        self.assertEqual(ret, [])

    def test_encoding_none_with_coding_comment(self):
        checker = CodingChecker(None, 'testsuite/utf8.py')
        checker.encodings = None
        ret = list(checker.run())
        self.assertEqual(len(ret), 1)
        self.assertEqual(ret[0][0], 1)
        self.assertEqual(ret[0][1], 0)
        self.assertTrue(ret[0][2].startswith('C103 '))


if __name__ == '__main__':
    unittest.main()


flake8_coding.py
# -*- coding: utf-8 -*-

import re

__version__ = '1.3.1'


class CodingChecker(object):
    name = 'flake8_coding'
    version = __version__

    def __init__(self, tree, filename):
        self.filename = filename

    @classmethod
    def add_options(cls, parser):
        parser.add_option(
            '--accept-encodings', default='latin-1, utf-8', action='store',
            help="Acceptable source code encodings for `coding:` magic comment"
        )
        parser.add_option(
            '--no-accept-encodings', action='store_true', parse_from_config=True,
            help="Warn for files containing a `coding:` magic comment"
        )

        if hasattr(parser, 'config_options'):  # for flake8 < 3.0
            parser.config_options.append('accept-encodings')
            parser.config_options.append('no-accept-encodings')

    @classmethod
    def parse_options(cls, options):
        if options.no_accept_encodings:
            cls.encodings = None
        else:
            cls.encodings = [e.strip().lower() for e in options.accept_encodings.split(',')]

    def read_headers(self):
        if self.filename in ('stdin', '-', None):
            try:
                # flake8 >= v3.0
                from flake8.engine import pep8 as stdin_utils
            except ImportError:
                from flake8 import utils as stdin_utils
            return stdin_utils.stdin_get_value().splitlines(True)[:2]
        else:
            try:
                import pycodestyle
            except ImportError:
                import pep8 as pycodestyle
            return pycodestyle.readlines(self.filename)[:2]

    def run(self):
        try:
            # PEP-263 says: a magic comment must be placed into the source
            #               files either as first or second line in the file
            lines = self.read_headers()
            if len(lines) == 0:
                return

            for lineno, line in enumerate(lines, start=1):
                matched = re.search(r'coding[:=]\s*([-\w.]+)', line, re.IGNORECASE)
                if matched:
                    if self.encodings:
                        if matched.group(1).lower() not in self.encodings:
                            yield lineno, 0, "C102 Unknown encoding found in coding magic comment", type(self)
                    else:
                        yield lineno, 0, "C103 Coding magic comment present", type(self)
                    break
            else:
                if self.encodings:
                    yield 1, 0, "C101 Coding magic comment not found", type(self)
        except IOError:
            pass
